src/Servers/Kestrel/test/InMemory.FunctionalTests/ResponseTests.cs

```diff
@@ -2554,21 +2554,24 @@ public async Task AppAbortIsLogged()
         }
 
         [Fact]
+        [Repeat(100)]
         [QuarantinedTest("https://github.com/dotnet/aspnetcore/issues/24164")]
         public async Task AppAbortViaIConnectionLifetimeFeatureIsLogged()
         {
             var testContext = new TestServiceContext(LoggerFactory);
 
-            await using (var server = new TestServer(httpContext =>
+            await using (var server = new TestServer(async httpContext =>
             {
                 var feature = httpContext.Features.Get<IConnectionLifetimeFeature>();
                 feature.Abort();
 
+                await Task.Delay(15);
```

Comment:
This is awful. It does stop failing locally with this though.

The issue I believe is because the other `ConnectionClosed` cts registration that cancels the HttpConnection can run before or after the below registration, and if it runs after then the response can be flushed before the abort.

I couldn't find a way to reliably wait for the other cts registration to run before we return from this func, so if anyone knows of a way that would be nice.

Reply:
I don't want to be merge arbitrary 15ms Task.Delays into our tests as that's likely to mask issues at best. What is the "other ConnectionClosed cts registration" you're referring to?

Reply:
https://github.com/dotnet/aspnetcore/blob/517c83ace0d96ce095a764f13d906016d2052283/src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs#L101
which will call into other methods and eventually set `_connectionAborted` to true which is checked before writing the output that causes this test to sometimes fail
https://github.com/dotnet/aspnetcore/blob/89c3d6bf23974b35ac896df92e9c588288599107/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs#L699

When I hit the issue under the debugger I saw that the above check had `_connectionAborted` set to true even though it was inside the if statement, so there is some race there which I assume is due to how the cancellation token callbacks run.

Reply:
Ok dug into this some more, and the issue is because the tests `Register` will always run before the OnConnectionClosed Register because cts's run in reverse registration order. So there is a race between ProduceEnd being called and the OnConnectionClosed callback running and setting the `_connectionAborted` bool.

We could technically add a callback during
https://github.com/dotnet/aspnetcore/blob/89c3d6bf23974b35ac896df92e9c588288599107/src/Servers/Kestrel/test/InMemory.FunctionalTests/TestTransport/TestServer.cs#L105-L110
That will run before `_transportFactory.AddConnection` so we have a chance to hook up a cts registration first.

Reply:
Hopefully last bit of digging: Relying on CTS ordering is flawed, the callbacks can be registered to random internal partitions based on the ID of the thread the register comes from.

I didn't notice before, but the InMemoryTransportConnection has a WaitForCloseTask that will complete after the cts is done canceling. I think we can switch to that to become reliable.